<a href="https://colab.research.google.com/github/markjschmidt/PET_Randoms/blob/main/AI_assisted_Random_Coincidence_Simulation_20240929.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Developing python code to create Monte Carlo simulations for random coincidence events generation.

Using Anthropic AI Claude 3 chat interface for coding assistance:



Provide python code that will run in Google Colab to demonstrate and plot Poisson distribution of Randoms coincidence events created by radioactive Fluorine-18 decay.

Using random number generation, simulate a F-18 source that produces 8 million decay events over 1 second, divisible by 6 nanosecond bins to simulate Philips Gemini coincidence timing window.

Zoom in to show a plotly scatter plot for 1 millisecond of the random data, 8000 decay events. The 8,000 events are then divided into 166,666 bins of 6 nanosecond sequential time slots.

Finally show a histogram of the Poisson distribution by X axis, number decay events that are in each bin and Y axis probability. The numbers will be small since there are only 8,000,000 decay events divided by 166,666,666 time slot bins. Print the number of bins that contain 2 events. These will be our simulated coincidence event pairs.




## Simulation of 8 million Singles Events

This code will generate two plots:

A scatter plot showing the random generated decay events within the first millisecond of the simulation. The x-axis represents time, and the y-axis represents the atom number that decayed.

A histogram representing the Poisson distribution of the decay events. The x-axis shows the number of decay events per 6 nanosecond bin, and the y-axis represents the probability. Additionally, it will print the number of bins that contain exactly two events.


### Mount Google Drive to save notebook:


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
%cd /content/drive/My\ Drive/Colab\ Notebooks


/content/drive/My Drive/Colab Notebooks


In [4]:
import matplotlib.pyplot as plt
import PIL.Image

# Import the display module from IPython
from IPython.display import display

In [ ]:
# Fetch a single <1MB file using the raw GitHub URL.
# !curl --remote-name \
#     -H 'Accept: application/vnd.github.v3.raw' \
#     --location https://api.github.com/repos/jakevdp/PythonDataScienceHandbook/contents/notebooks/data/california_cities.csv


## Start of Simulation Code

In [6]:
import numpy as np
import plotly.graph_objs as go
from scipy.stats import poisson

# Simulate a F-18 source that produces 8,000,000 decay events over 1 second,
# visualize results scaled to 8,000 / 1 millisecond
# with a time resolution of 6 nanoseconds
total_events = 8000
time_resolution = 6e-9  # 6 nanoseconds
total_time = 1e-3  # 1 millisecond
event_times = np.random.uniform(0, total_time, total_events)

# Scatter plot of the 8,000 random simulated events
trace1 = go.Scatter(x=event_times,
                    y=np.arange(1, total_events + 1),
                    mode='markers',
                    marker=dict(color='blue'))

# Layout and plot for trace1
layout1 = go.Layout(title='Scatter Plot of Fluorine-18 Decay Events',
                    xaxis=dict(title='Time (s)', range=[0, total_time]),
                    yaxis=dict(title='Atom Number', range=[0, total_events + 1]))
fig1 = go.Figure(data=[trace1], layout=layout1)
fig1.show()

# Divide the 8,000 events into 166,666 bins of 6 nanosecond sequential time slots
num_bins = int(total_time // time_resolution)
event_counts, _ = np.histogram(event_times, bins=num_bins, range=(0, total_time))

# Calculate Poisson probabilities for bin counts from 0 to 4
mu = total_events / num_bins  # Mean of the Poisson distribution
poisson_probs = [poisson.pmf(k, mu) for k in range(5)]

#This line calculates the Poisson probabilities for bin counts ranging from 0 to 4 (inclusive) using a list comprehension.
#poisson.pmf is a function from the scipy.stats module that computes the probability mass function (PMF) of the Poisson distribution.
#The pmf function takes two arguments: k (the number of events) and mu (the mean of the distribution).
#The list comprehension [poisson.pmf(k, mu) for k in range(5)] iterates over k values from 0 to 4 and calculates the Poisson PMF for each value of k using the previously calculated mu.
#The resulting list poisson_probs will contain the Poisson probabilities for bin counts 0, 1, 2, 3, and 4, given the calculated mean mu.

# Note that the Poisson distribution is discrete, and the poisson.pmf function calculates the probability mass for specific integer values (k).
# The returned value is also a numpy array. The definition of the Poisson pmf is poisson.pmf(k) = exp(-mu) * mu^k / k! for k >= 0 and mu >= 0.

# Create a bar plot for the Poisson distribution
trace2 = go.Bar(x=np.arange(5),
                y=poisson_probs,
                marker_color='red',
                opacity=0.75)

# Print the number of bins that contain 2 events
num_bins_with_2_events = np.sum(event_counts == 2)
print(f"Number of bins containing 2 events: {num_bins_with_2_events}")

# Print the probability of 2 decay events per bin
prob_2_events = poisson.pmf(2, mu)
print(f"Probability of 2 decay events per bin: {prob_2_events:.6f}")

# Layout and plot for trace2
layout2 = go.Layout(title='Poisson Distribution of Fluorine-18 Decay Events',
                    xaxis=dict(title='Number of Decay Events per Bin', tickvals=[0, 1, 2, 3, 4]),
                    yaxis=dict(title='Probability'))
fig2 = go.Figure(data=[trace2], layout=layout2)
fig2.show()


Number of bins containing 2 events: 165
Probability of 2 decay events per bin: 0.001098


Some code detailed explanation:

mu = total_events / num_bins # Mean of the Poisson distribution
poisson_probs = [poisson.pmf(k, mu) for k in range(5)]


mu = total_events / num_bins:

This line calculates the mean (mu) of the Poisson distribution.
The mean is calculated by dividing total_events by num_bins.
total_events is likely a variable representing the total number of events or occurrences being modeled.

num_bins is likely a variable representing the number of bins or intervals into which the events are divided.

By dividing total_events by num_bins, we obtain the average number of events per bin, which is the mean (mu) of the Poisson distribution.



poisson_probs = [poisson.pmf(k, mu) for k in range(5)]:

This line calculates the Poisson probabilities for bin counts ranging from 0 to 4 (inclusive) using a list comprehension.

poisson.pmf is a function from the scipy.stats module that computes the probability mass function (PMF) of the Poisson distribution.
The pmf function takes two arguments: k (the number of events) and mu (the mean of the distribution).





The list comprehension [poisson.pmf(k, mu) for k in range(5)] iterates over k values from 0 to 4 and calculates the Poisson PMF for each value of k using the previously calculated mu.
The resulting list poisson_probs will contain the Poisson probabilities for bin counts 0, 1, 2, 3, and 4, given the calculated mean mu.

In the context of the Poisson distribution, these lines of code are likely used to calculate the probabilities of observing different bin counts (0, 1, 2, 3, and 4) based on the given mean (mu). The Poisson distribution is often used to model the probability of a given number of events occurring in a fixed interval of time or space, assuming that the events occur independently and at a constant average rate (mu).
For example, if you are modeling the number of radioactive decays in a fixed time interval, total_events could represent the total number of decays observed, num_bins could represent the number of time intervals, and mu would be the average number of decays per time interval. The poisson_probs list would then contain the probabilities of observing 0, 1, 2, 3, and 4 decays in a single time interval, given the calculated mean (mu).
Note that the Poisson distribution is discrete, and the poisson.pmf function calculates the probability mass for specific integer values (k). If you need to calculate probabilities for non-integer values or continuous ranges, you would need to use a different distribution or approach.








Note: Due to only 8,000,000 decay events divided in 166,666,666 time slot bins in 1 second or 8,000 events divided in 166,666 bins for a 1 msec sample, over 95% of the bins have 0 events.

The event decay times is the random value generated. Since they are random values, not evenly spaced clocked values, there will be some small number of events occuring within the same time bins. This simulated data set resulted in approximately 166 bins with 2 events.

The Poisson Distribution accurately calculated the number at 0.001098 of 8,000,000 events = 183




### Increase singles events by 4x to 32,000 events in 1 millisecond

In [7]:
import numpy as np
import plotly.graph_objs as go
from scipy.stats import poisson

# Simulate a F-18 source that produces 8,000,000 decay events over 1 second,
# visualize results scaled to 8,000 / 1 millisecond
# with a time resolution of 6 nanoseconds
total_events = 32000
time_resolution = 6e-9  # 6 nanoseconds
total_time = 1e-3  # 1 millisecond
event_times = np.random.uniform(0, total_time, total_events)

# Scatter plot of the 8,000 random simulated events
trace1 = go.Scatter(x=event_times,
                    y=np.arange(1, total_events + 1),
                    mode='markers',
                    marker=dict(color='blue'))

# Layout and plot for trace1
layout1 = go.Layout(title='Scatter Plot of Fluorine-18 Decay Events',
                    xaxis=dict(title='Time (s)', range=[0, total_time]),
                    yaxis=dict(title='Atom Number', range=[0, total_events + 1]))
fig1 = go.Figure(data=[trace1], layout=layout1)
fig1.show()

# Divide the 8,000 events into 166,666 bins of 6 nanosecond sequential time slots
num_bins = int(total_time // time_resolution)
event_counts, _ = np.histogram(event_times, bins=num_bins, range=(0, total_time))

# Calculate Poisson probabilities for bin counts from 0 to 4
mu = total_events / num_bins  # Mean of the Poisson distribution
poisson_probs = [poisson.pmf(k, mu) for k in range(5)]

# Create a bar plot for the Poisson distribution
trace2 = go.Bar(x=np.arange(5),
                y=poisson_probs,
                marker_color='red',
                opacity=0.75)

# Print the number of bins that contain 2 events
num_bins_with_2_events = np.sum(event_counts == 2)
print(f"Number of bins containing 2 events: {num_bins_with_2_events}")

# Print the probability of 2 decay events per bin
prob_2_events = poisson.pmf(2, mu)
print(f"Probability of 2 decay events per bin: {prob_2_events:.6f}")

# Layout and plot for trace2
layout2 = go.Layout(title='Poisson Distribution of Fluorine-18 Decay Events',
                    xaxis=dict(title='Number of Decay Events per Bin', tickvals=[0, 1, 2, 3, 4]),
                    yaxis=dict(title='Probability'))
fig2 = go.Figure(data=[trace2], layout=layout2)
fig2.show()


Number of bins containing 2 events: 2548
Probability of 2 decay events per bin: 0.015212


### Now with 4x the singles events we have approximately 15x the coincidence events.

### Very good! Let's review our terminology because I believe it is affecting our printed numerical results. Events refers to single randomly generated radioactive decays. These can also be referred to as 'singles'.
### When a large number of randomly generated events are occurring close enough in time, two or more can occur in 'coincidence' and will be stored in the same 6 nanosecond bin. A 'coincidence event' therefore is a binned pair of events and is only counted as 1 coincidence event.

A AI:

You're absolutely right, thank you for clarifying the terminology. I understand now:

"Events" or "singles" refer to individual randomly generated radioactive decays.
When two or more events/singles occur within the same 6 nanosecond time bin, they are considered a "coincidence event".
A "coincidence event" counts as a single binned pair of events, even if more than two events/singles fall within the same 6 nanosecond bin.

With this clarification, let's update the code to accurately count and report the "coincidence events" and the total number of "events" or "singles":

The main changes are:

The num_coincidence_events is now calculated as len(coincidence_bin_ranges), which gives the number of bins containing two or more events (coincidence events).
The y-axis title for the scatter plots is changed to "Event Number" to be more accurate.
The probability printed is now prob_coincidence_event, which is calculated as 1 - poisson.pmf(0, mu) - poisson.pmf(1, mu), representing the probability of having two or more events in a bin (a coincidence event).

With these changes, the code should accurately report the number of "coincidence events" (binned pairs of events) and the total number of "events" or "singles" generated. The scatter plots will also use the correct terminology in the axis labels.



### 1 millisecond,  8000 events

In [8]:
import numpy as np
import plotly.graph_objs as go
from scipy.stats import poisson

# Simulate a F-18 source that produces 8,000 decay events over 1 millisecond
# with a time resolution of 6 nanoseconds
total_events = 8000
time_resolution = 6e-9  # 6 nanoseconds
total_time = 10e-4  # 1 millisecond
event_times = np.random.uniform(0, total_time, total_events)

# Divide the 8,000 events into 166,666 bins of 6 nanosecond sequential time slots
num_bins = int(total_time // time_resolution)
event_counts, bin_edges = np.histogram(event_times, bins=num_bins, range=(0, total_time))

# Get the indices of bins containing 2 events
coincidence_bin_indices = np.where(event_counts == 2)[0]
coincidence_bin_ranges = [(bin_edges[i], bin_edges[i + 1]) for i in coincidence_bin_indices]

# Scatter plot of the 8,000 random simulated events
marker_colors = []
for t in event_times:
    for bin_range in coincidence_bin_ranges:
        if bin_range[0] <= t < bin_range[1]:
            marker_colors.append('red')
            break
    else:
        marker_colors.append('blue')

trace1 = go.Scatter(x=event_times,
                    y=np.arange(1, total_events + 1),
                    mode='markers',
                    marker=dict(color=marker_colors))


# Layout and plot for trace1
layout1 = go.Layout(title='Scatter Plot of Fluorine-18 Decay Events',
                    xaxis=dict(title='Time (s)', range=[0, total_time]),
                    yaxis=dict(title='Atom Number', range=[0, total_events + 1]))
fig1 = go.Figure(data=[trace1], layout=layout1)
fig1.show()

# Calculate Poisson probabilities for bin counts from 1 to 4
mu = total_events / num_bins  # Mean of the Poisson distribution
poisson_probs = [poisson.pmf(k, mu) for k in range(1, 5)]

# Create a bar plot for the Poisson distribution
trace2 = go.Bar(x=np.arange(1, 5),
                y=poisson_probs,
                marker_color='red',
                opacity=0.75)

# Print the number of bins that contain 2 events
num_bins_with_2_events = np.sum(event_counts == 2)
print(f"Number of bins containing 2 events: {num_bins_with_2_events}")

# Count the red markers (coincidence events)
num_coincidence_events = marker_colors.count('red')
print(f"Number of coincidence events: {num_coincidence_events}")


# Print the probability of 2 decay events per bin
prob_2_events = poisson.pmf(2, mu)
print(f"Probability of 2 decay events per bin: {prob_2_events:.6f}")

# Layout and plot for trace2
layout2 = go.Layout(title='Poisson Distribution of Fluorine-18 Decay Events',
                    xaxis=dict(title='Number of Decay Events per Bin', tickvals=[1, 2, 3, 4]),
                    yaxis=dict(title='Probability'))
fig2 = go.Figure(data=[trace2], layout=layout2)
fig2.show()


Number of bins containing 2 events: 198
Number of coincidence events: 396
Probability of 2 decay events per bin: 0.001098


### 100 microseconds,  800 events

In [9]:
import numpy as np
import plotly.graph_objs as go
from scipy.stats import poisson

# Simulate a F-18 source that produces 8,000 decay events over 1 millisecond
# with a time resolution of 6 nanoseconds
total_events = 800
time_resolution = 6e-9  # 6 nanoseconds
total_time = 10e-5  # 1 millisecond
event_times = np.random.uniform(0, total_time, total_events)

# Divide the 8,000 events into 166,666 bins of 6 nanosecond sequential time slots
num_bins = int(total_time // time_resolution)
event_counts, bin_edges = np.histogram(event_times, bins=num_bins, range=(0, total_time))

# Get the indices of bins containing 2 events
coincidence_bin_indices = np.where(event_counts == 2)[0]
coincidence_bin_ranges = [(bin_edges[i], bin_edges[i + 1]) for i in coincidence_bin_indices]

# Scatter plot of the 8,000 random simulated events
marker_colors = []
for t in event_times:
    for bin_range in coincidence_bin_ranges:
        if bin_range[0] <= t < bin_range[1]:
            marker_colors.append('red')
            break
    else:
        marker_colors.append('blue')

trace1 = go.Scatter(x=event_times,
                    y=np.arange(1, total_events + 1),
                    mode='markers',
                    marker=dict(color=marker_colors))


# Layout and plot for trace1
layout1 = go.Layout(title='Scatter Plot of Fluorine-18 Decay Events',
                    xaxis=dict(title='Time (s)', range=[0, total_time]),
                    yaxis=dict(title='Atom Number', range=[0, total_events + 1]))
fig1 = go.Figure(data=[trace1], layout=layout1)
fig1.show()

# Calculate Poisson probabilities for bin counts from 1 to 4
mu = total_events / num_bins  # Mean of the Poisson distribution
poisson_probs = [poisson.pmf(k, mu) for k in range(1, 5)]

# Create a bar plot for the Poisson distribution
trace2 = go.Bar(x=np.arange(1, 5),
                y=poisson_probs,
                marker_color='red',
                opacity=0.75)

# Print the number of bins that contain 2 events
num_bins_with_2_events = np.sum(event_counts == 2)
print(f"Number of bins containing 2 events: {num_bins_with_2_events}")

# Count the red markers (coincidence events)
num_coincidence_events = marker_colors.count('red')
print(f"Number of coincidence events: {num_coincidence_events}")


# Print the probability of 2 decay events per bin
prob_2_events = poisson.pmf(2, mu)
print(f"Probability of 2 decay events per bin: {prob_2_events:.6f}")

# Layout and plot for trace2
layout2 = go.Layout(title='Poisson Distribution of Fluorine-18 Decay Events',
                    xaxis=dict(title='Number of Decay Events per Bin', tickvals=[1, 2, 3, 4]),
                    yaxis=dict(title='Probability'))
fig2 = go.Figure(data=[trace2], layout=layout2)
fig2.show()


Number of bins containing 2 events: 17
Number of coincidence events: 34
Probability of 2 decay events per bin: 0.001098


### Zoom in to 10 microsecond window with 80 total decay events.

In [10]:
import numpy as np
import plotly.graph_objs as go
from scipy.stats import poisson

# Simulate a F-18 source that produces 8,000 decay events over 1 millisecond
# with a time resolution of 6 nanoseconds
total_events = 80
time_resolution = 6e-9  # 6 nanoseconds
total_time = 10e-6  # 10 microsecond
event_times = np.random.uniform(0, total_time, total_events)

# Divide the 8,000 events into 166,666 bins of 6 nanosecond sequential time slots
num_bins = int(total_time // time_resolution)
event_counts, bin_edges = np.histogram(event_times, bins=num_bins, range=(0, total_time))

# Get the indices of bins containing 2 events
coincidence_bin_indices = np.where(event_counts == 2)[0]
coincidence_bin_ranges = [(bin_edges[i], bin_edges[i + 1]) for i in coincidence_bin_indices]

# Scatter plot of the 8,000 random simulated events
marker_colors = []
for t in event_times:
    for bin_range in coincidence_bin_ranges:
        if bin_range[0] <= t < bin_range[1]:
            marker_colors.append('red')
            break
    else:
        marker_colors.append('blue')

trace1 = go.Scatter(x=event_times,
                    y=np.arange(1, total_events + 1),
                    mode='markers',
                    marker=dict(color=marker_colors))


# Layout and plot for trace1
layout1 = go.Layout(title='Scatter Plot of Fluorine-18 Decay Events',
                    xaxis=dict(title='Time (s)', range=[0, total_time]),
                    yaxis=dict(title='Atom Number', range=[0, total_events + 1]))
fig1 = go.Figure(data=[trace1], layout=layout1)
fig1.show()

# Calculate Poisson probabilities for bin counts from 1 to 4
mu = total_events / num_bins  # Mean of the Poisson distribution
poisson_probs = [poisson.pmf(k, mu) for k in range(1, 5)]

# Create a bar plot for the Poisson distribution
trace2 = go.Bar(x=np.arange(1, 5),
                y=poisson_probs,
                marker_color='red',
                opacity=0.75)

# Print the number of bins that contain 2 events
num_bins_with_2_events = np.sum(event_counts == 2)
print(f"Number of bins containing 2 events: {num_bins_with_2_events}")

# Count the red markers (coincidence events)
num_coincidence_events = marker_colors.count('red')
print(f"Number of coincidence events: {num_coincidence_events}")


# Print the probability of 2 decay events per bin
prob_2_events = poisson.pmf(2, mu)
print(f"Probability of 2 decay events per bin: {prob_2_events:.6f}")

# Layout and plot for trace2
layout2 = go.Layout(title='Poisson Distribution of Fluorine-18 Decay Events',
                    xaxis=dict(title='Number of Decay Events per Bin', tickvals=[1, 2, 3, 4]),
                    yaxis=dict(title='Probability'))
fig2 = go.Figure(data=[trace2], layout=layout2)
fig2.show()


Number of bins containing 2 events: 2
Number of coincidence events: 4
Probability of 2 decay events per bin: 0.001099


### Zoom in to 10 microsecond window with 80 total decay events.

### Highlight Coincidence events in Red.

### Add another plot. The second shows atoms/ decay events sorted by decay time to see coincidence events (Red) more clearly.

### Remove histogram bar for 0 event bins so small 2 event bin is clearly seen.


In [11]:
import numpy as np
import plotly.graph_objs as go
from scipy.stats import poisson

# Simulate a F-18 source that produces 80 events (singles) over 10 microseconds
# with a time resolution of 6 nanoseconds
total_events = 80
time_resolution = 6e-9  # 6 nanoseconds
total_time = 10e-6  # 10 microseconds

event_times = np.random.uniform(0, total_time, total_events)

# Divide the 80 events into 1,666 bins of 6 nanosecond sequential time slots
num_bins = int(total_time // time_resolution)
event_counts, bin_edges = np.histogram(event_times, bins=num_bins, range=(0, total_time))

# Get the indices of bins containing 2 or more events (coincidence events)
coincidence_bin_indices = np.where(event_counts >= 2)[0]
coincidence_bin_ranges = [(bin_edges[i], bin_edges[i + 1]) for i in coincidence_bin_indices]

# Scatter plot of the 80 random simulated events
marker_colors = []
coincidence_event_times = []

for t in event_times:
    for bin_range in coincidence_bin_ranges:
        if bin_range[0] <= t < bin_range[1]:
            marker_colors.append('red')
            coincidence_event_times.append(t)
            break
    else:
        marker_colors.append('blue')

trace1 = go.Scatter(x=event_times,
                    y=np.arange(1, total_events + 1),
                    mode='markers',
                    marker=dict(color=marker_colors))

# Count the number of coincidence events
num_coincidence_events = len(coincidence_bin_ranges)
print(f"Number of coincidence events: {num_coincidence_events}")

# Layout and plot for trace1
layout1 = go.Layout(title='Scatter Plot of Fluorine-18 Decay Events',
                    xaxis=dict(title='Time (s)', range=[0, total_time]),
                    yaxis=dict(title='Event Number', range=[0, total_events + 1]))
fig1 = go.Figure(data=[trace1], layout=layout1)
fig1.show()

# Sort event times and create a new scatter plot
sorted_event_times = sorted(event_times)
sorted_event_numbers = np.arange(1, total_events + 1)

# Handle coincidence events by plotting them slightly above/below each other
coincidence_event_idx = []
trace4_x, trace4_y = [], []
for i, t in enumerate(sorted_event_times):
    for bin_range in coincidence_bin_ranges:
        if bin_range[0] <= t < bin_range[1]:
            coincidence_event_idx.append(i)
            trace4_x.extend([t, t])
            trace4_y.extend([sorted_event_numbers[i], sorted_event_numbers[i] + 0.1])
            break
    else:
        trace4_x.append(t)
        trace4_y.append(sorted_event_numbers[i])

# Color the markers based on coincidence events
trace4_marker_colors = ['red' if i in coincidence_event_idx else 'blue' for i in range(total_events)]

trace4 = go.Scatter(x=trace4_x,
                    y=trace4_y,
                    mode='markers',
                    marker=dict(color=trace4_marker_colors))

# Layout and plot for trace4
layout4 = go.Layout(title='Scatter Plot of All Events (Coincidence in Red)',
                    xaxis=dict(title='Time (s)', range=[0, total_time]),
                    yaxis=dict(title='Event Number', range=[1, total_events + 1]))
fig4 = go.Figure(data=[trace4], layout=layout4)
fig4.show()

# Calculate Poisson probabilities for bin counts from 1 to 4
mu = total_events / num_bins  # Mean of the Poisson distribution
poisson_probs = [poisson.pmf(k, mu) for k in range(1, 5)]

# Create a bar plot for the Poisson distribution
trace2 = go.Bar(x=np.arange(1, 5),
                y=poisson_probs,
                marker_color='red',
                opacity=0.75)

# Print the probability of 2 or more events per bin (coincidence event)
prob_coincidence_event = 1 - poisson.pmf(0, mu) - poisson.pmf(1, mu)
print(f"Probability of a coincidence event per bin: {prob_coincidence_event:.6f}")

# Layout and plot for trace2
layout2 = go.Layout(title='Poisson Distribution of Fluorine-18 Decay Events',
                    xaxis=dict(title='Number of Events per Bin', tickvals=[1, 2, 3, 4]),
                    yaxis=dict(title='Probability'))
fig2 = go.Figure(data=[trace2], layout=layout2)
fig2.show()


Number of coincidence events: 2


Probability of a coincidence event per bin: 0.001117


Additional code explanation:

Here's what we've changed in the code:

We added event_counts, bin_edges = np.histogram(event_times, bins=num_bins, range=(0, total_time)) to get the bin edges in addition to the event counts.
We find the indices of bins containing 2 events using coincidence_bin_indices = np.where(event_counts == 2)[0].
We create a list of tuples coincidence_bin_ranges containing the bin ranges for bins with 2 events.
We create a list marker_colors to store the marker color for each event. If the event time falls within a bin range with 2 events, we set the marker color to 'red', otherwise 'blue'.
We pass the marker_colors list to the marker argument of go.Scatter for trace1.

Now, when you run this code, the scatter plot (trace1) will display events in bins containing 2 events (coincidence events) with red markers, and all other events with blue markers.
The rest of the code remains the same, calculating and displaying the Poisson distribution probabilities for bin counts from 0 to 4, and printing the number of bins containing 2 events and the probability of 2 decay events per bin.



### Diagnostic to run 10 trials of simulated events


### Here's the updated code that prompts the user for both the total number of events and the total time in microseconds:


### Note: Don't enter 8000000 for events & 1000000 for nano seconds. Consumes to much compute resources and won't finish.

### Example use: 80 events in 10 microseconds; 800 events in 100 microseconds; 8000 events in 1000 microseconds (1 millisecond).




In [12]:
import numpy as np

# Prompt user for total number of events and total time
total_events = int(input("Enter the total number of events: "))
total_time_microseconds = float(input("Enter the total time in microseconds: "))
total_time = total_time_microseconds * 1e-6  # Convert to seconds

# Run the simulation multiple times and count the number of coincidence events
num_runs = 10

total_coincidence_events = 0

for i in range(num_runs):
    # Simulate a F-18 source that produces the specified number of decay events over the specified time
    # with a time resolution of 6 nanoseconds
    time_resolution = 6e-9  # 6 nanoseconds
    event_times = np.random.uniform(0, total_time, total_events)

    # Divide the events into bins of 6 nanosecond sequential time slots
    num_bins = int(total_time // time_resolution)
    event_counts, bin_edges = np.histogram(event_times, bins=num_bins, range=(0, total_time))

    # Get the indices of bins containing 2 events
    coincidence_bin_indices = np.where(event_counts == 2)[0]
    coincidence_bin_ranges = [(bin_edges[i], bin_edges[i + 1]) for i in coincidence_bin_indices]

    # Count the coincidence events
    num_coincidence_events = len(coincidence_bin_ranges)
    total_coincidence_events += num_coincidence_events

    print(f"Run {i+1}: Number of coincidence events = {num_coincidence_events}")

print(f"\nAverage number of coincidence events over {num_runs} runs: {total_coincidence_events / num_runs:.2f}")


Enter the total number of events: 8000
Enter the total time in microseconds: 1000
Run 1: Number of coincidence events = 213
Run 2: Number of coincidence events = 180
Run 3: Number of coincidence events = 194
Run 4: Number of coincidence events = 175
Run 5: Number of coincidence events = 190
Run 6: Number of coincidence events = 201
Run 7: Number of coincidence events = 168
Run 8: Number of coincidence events = 180
Run 9: Number of coincidence events = 179
Run 10: Number of coincidence events = 177

Average number of coincidence events over 10 runs: 185.70


Me:

Would you please recall back to the code you provided to me here in this chat. Can you provide a detailed explanation for this line of code:

event_counts, bin_edges = np.histogram(event_times, bins=num_bins, range=(0, total_time))

Does using bin_edges create the 6 nanosecond wide windows or bins used to determine if 2 events are in coincidence?


AI:

Yes, the bin_edges created by the np.histogram function is what defines the boundaries of the 6 nanosecond wide bins or windows used to determine if two events are in coincidence.
Let's break down the line of code:

np.histogram is a NumPy function that computes the histogram of a set of data points.
event_times is the array of randomly generated event times (radioactive decays) in the time range [0, total_time].
bins=num_bins specifies the number of bins or intervals into which the event_times should be divided. In this case, num_bins is calculated as int(total_time // time_resolution), which ensures that each bin has a width of time_resolution (6 nanoseconds).
range=(0, total_time) specifies the range of the histogram, which is the same as the range of event_times.
event_counts, bin_edges = ... unpacks the output of np.histogram into two variables:

event_counts is an array containing the number of events (radioactive decays) in each bin.
bin_edges is an array containing the boundaries of the bins.



The bin_edges array contains the left and right edges of each bin. For example, if bin_edges is [0.0, 6e-9, 12e-9, ...], then the first bin is [0.0, 6e-9), the second bin is [6e-9, 12e-9), and so on. The width of each bin is determined by the difference between consecutive bin edges, which is equal to time_resolution (6 nanoseconds).
So, when you later check for event_counts == 2 and use the corresponding bin_edges indices, you are effectively identifying the bins (6 nanosecond windows) where two events occurred in coincidence.
In summary, yes, the bin_edges array created by np.histogram is what defines the 6 nanosecond wide windows or bins used to determine if two events are in coincidence.


NOTE: For 1 msec we are looking at 166,667 bins that contain 180 mean coincidence pairs which is 0.001 of the bins measured.



Try our min rate of Singles: 1,300,000 or 1,300 in 1 msec;  
Look at at 6 nsec. and 4 nsec.

In [13]:
import numpy as np

# Prompt user for total number of events and total time
total_events = int(input("Enter the total number of events: "))
total_time_microseconds = float(input("Enter the total time in microseconds: "))
total_time = total_time_microseconds * 1e-6  # Convert to seconds

# Run the simulation multiple times and count the number of coincidence events
num_runs = 10

total_coincidence_events = 0

for i in range(num_runs):
    # Simulate a F-18 source that produces the specified number of decay events over the specified time
    # with a time resolution of 6 nanoseconds
    time_resolution = 6e-9  # 6 nanoseconds
    event_times = np.random.uniform(0, total_time, total_events)

    # Divide the events into bins of 6 nanosecond sequential time slots
    num_bins = int(total_time // time_resolution)
    event_counts, bin_edges = np.histogram(event_times, bins=num_bins, range=(0, total_time))

    # Get the indices of bins containing 2 events
    coincidence_bin_indices = np.where(event_counts == 2)[0]
    coincidence_bin_ranges = [(bin_edges[i], bin_edges[i + 1]) for i in coincidence_bin_indices]

    # Count the coincidence events
    num_coincidence_events = len(coincidence_bin_ranges)
    total_coincidence_events += num_coincidence_events

    print(f"Run {i+1}: Number of coincidence events = {num_coincidence_events}")

print(f"\nAverage number of coincidence events over {num_runs} runs: {total_coincidence_events / num_runs:.2f}")


Enter the total number of events: 1300
Enter the total time in microseconds: 1000
Run 1: Number of coincidence events = 2
Run 2: Number of coincidence events = 5
Run 3: Number of coincidence events = 2
Run 4: Number of coincidence events = 3
Run 5: Number of coincidence events = 7
Run 6: Number of coincidence events = 4
Run 7: Number of coincidence events = 7
Run 8: Number of coincidence events = 4
Run 9: Number of coincidence events = 3
Run 10: Number of coincidence events = 3

Average number of coincidence events over 10 runs: 4.00


Then 4 nsec.

In [14]:
import numpy as np

# Prompt user for total number of events and total time
total_events = int(input("Enter the total number of events: "))
total_time_microseconds = float(input("Enter the total time in microseconds: "))
total_time = total_time_microseconds * 1e-6  # Convert to seconds

# Run the simulation multiple times and count the number of coincidence events
num_runs = 10

total_coincidence_events = 0

for i in range(num_runs):
    # Simulate a F-18 source that produces the specified number of decay events over the specified time
    # with a time resolution of 6 nanoseconds
    time_resolution = 4e-9  # 4 nanoseconds
    event_times = np.random.uniform(0, total_time, total_events)

    # Divide the events into bins of 6 nanosecond sequential time slots
    num_bins = int(total_time // time_resolution)
    event_counts, bin_edges = np.histogram(event_times, bins=num_bins, range=(0, total_time))

    # Get the indices of bins containing 2 events
    coincidence_bin_indices = np.where(event_counts == 2)[0]
    coincidence_bin_ranges = [(bin_edges[i], bin_edges[i + 1]) for i in coincidence_bin_indices]

    # Count the coincidence events
    num_coincidence_events = len(coincidence_bin_ranges)
    total_coincidence_events += num_coincidence_events

    print(f"Run {i+1}: Number of coincidence events = {num_coincidence_events}")

print(f"\nAverage number of coincidence events over {num_runs} runs: {total_coincidence_events / num_runs:.2f}")


Enter the total number of events: 8000
Enter the total time in microseconds: 1000
Run 1: Number of coincidence events = 109
Run 2: Number of coincidence events = 148
Run 3: Number of coincidence events = 116
Run 4: Number of coincidence events = 130
Run 5: Number of coincidence events = 104
Run 6: Number of coincidence events = 115
Run 7: Number of coincidence events = 127
Run 8: Number of coincidence events = 114
Run 9: Number of coincidence events = 113
Run 10: Number of coincidence events = 125

Average number of coincidence events over 10 runs: 120.10
